cam/cam.go

```diff
@@ -270,16 +269,36 @@ func (vs *videostore) DoCommand(_ context.Context, command map[string]interface{
 	// The response contains the name of the uploaded file.
 	case "save":
 		vs.logger.Debug("save command received")
-		from, to, metadata, err := validateSaveCommand(command)
+		from, to, metadata, async, err := validateSaveCommand(command)
 		if err != nil {
 			return nil, err
 		}
-		uploadFilePath, err := vs.conc.concat(from, to, metadata, vs.uploadPath)
+		uploadFilePath := generateOutputFilePath(vs.name.Name, formatDateTimeToString(from), metadata, vs.uploadPath)
+		uploadFileName := filepath.Base(uploadFilePath)
+		// Async save command will run the concat operation in the background.
+		// It waits for the segment duration before running to ensure the last segment
+		// is written to storage before concatenation.
+		// TODO: (seanp) Optimize this to immediately run as soon as the current segment is completed.
+		if async {
+			vs.logger.Debug("running save command asynchronously")
+			go func() {
+				time.Sleep(vs.conc.segmentDur)
+				err = vs.conc.concat(from, to, uploadFilePath)
+				if err != nil {
+					vs.logger.Error("failed to concat files ", err)
+				}
+			}()
```

Comment:
Good call out. Does this make sense to have the routine be preemtible through DoCommand context?
Something like:

```
            go func(ctx context.Context) {
                select {
                case <-time.After(vs.conc.segmentDur):
                    err = vs.conc.concat(from, to, uploadFilePath)
                    if err != nil {
                        vs.logger.Error("failed to concat files ", err)
                    }
                case <-ctx.Done():
                    vs.logger.Error("concat operation canceled")
                }
            }(ctx)
 ```
 
 Trying to figure out if this even makes sense... Would the camera client be able to cancel through the context as such?

Reply:
> Do we have a deadline or a context cancellation case where we want to stop this is if it too long running? Imagine i mistype in a request for hours of video concatentation by mistake, is there a way to save me from this go routine?

I can for sure add a common sense timeout here.

> On the other hand, if I spit many DoCmmands in because I spam the mobile app button, is there a way to ensure I down' spawn thousands of goroutines?

Would like to avoid over-engineering a manager solution here to keep track of open async requests to prevent this. But if we want to, I could create a ticket for this case. This is sort of already a vulnerability already to any of the API endpoints, but agree it is more problematic in the async case.

> Could we even not return from vs.conc.concat?

Do not forsee any hanging issues from concat ffmpeg calls, but a reasonable timeout should handle such a case.

Reply:
> the routine be preemtible through DoCommand context?

yes for go since the user could set a `context.WithDeadline`, but a DoCommand coming from Flutter wouldn't have the same context input I think. We can add a background context that we have complete control over, and error out if the stream timespan is larger than X hours or something, this can change.

> Would like to avoid over-engineering a manager solution here to keep track of open async requests to prevent this.

We maybe should add a stop/cancel DoCommand in case someone needs to bail and this is a problem for users of the module. We can do nothing now.

Reply:
> yes for go since the user could set a context.WithDeadline, but a DoCommand coming from Flutter wouldn't have the same context

Makes sense. I am not checking if context is done in the async routine which should hopefully handle that from the go case.

Also added a failsafe timeout.